### PR TITLE
Reload squashed files

### PR DESCRIPTION
**Pull request by TomNicholas**
Additional option for loading from files which were previously squashed using `ds.bout.save(separate_vars=True)`.

### CODE REVIEW

**Comment by TomNicholas** [comment]
Yes it is - does collect just save the most recent value of iteration? If so it has to be special-cased, which is what I've done in this most recent commit.

**Comment by TomNicholas** [comment]
Actually wait no that was not the right thing to do - I should have changed `xbout.collect` to match `boutdata.collect`

File: xbout/load.py [hunk 1]
```diff
@@ -123,6 +123,10 @@ def open_boutdataset(datapath='./BOUT.dmp.*.nc', inputfilepath=None,
         # bool attributes
         metadata['keep_xboundaries'] = int(keep_xboundaries)
         metadata['keep_yboundaries'] = int(keep_yboundaries)
+        if 'iteration' in ds:
```
Comment (johnomotani) [inline xbout/load.py]: I think this needs a `for variable in _BOUT_TIME_DEPENDENT_META_VARS` loop?

**Comment by johnomotani** [comment]
`boutdata.collect` doesn't handle sets of files split up along `t`, so doesn't see the issue with `iteration` (I think it also only reads scalars from `BOUT.dmp.0.nc` without checking if they're consistent).

Taking `iteration` from the most recent seems like a reasonable thing to do, and a useful pattern in case there are any other variables that somehow record the 'final state' of something in the dump files. :+1:

**Comment by johnomotani** [comment]
> Actually wait no that was not the right thing to do - I should have changed xbout.collect to match boutdata.collect, not changed xbout.open_boutdataset to match boutdata.collect.

I think the root of the problem was inconsistent values of `iteration` between different files, which caused an exception rather than an incorrect result making the test fail, so I think updating `open_boutdataset` is the correct fix.

**Comment by TomNicholas** [comment]
Okay I've made `open_boutdataset` return the most recent value of iteration from the `BOUT.dmp.0.nc` files.

`xbout.collect` now does the same thing, should it guarantee to return a scalar instead?

**Comment by johnomotani** [comment]
> xbout.collect now does the same thing, should it guarantee to return a scalar instead?

Isn't `iteration` a scalar already?

**Comment by TomNicholas** [comment]
> Isn't iteration a scalar already?

Not in general, that's what makes this complicated. Iteration is a per-processor variable, but it also depends on time. If you concatenate in time, it makes sense to end up with a 1D array along `'t'` which has a different value depending on which job out of the overall run that was. 

In `xbout.collect` it seems to return a differently-shaped array depending on what you're concatenating from, I'm asking whether we should just always force it to be a scalar (picking the last value from the array `open_boutdataset` returns)?

**Comment by johnomotani** [comment]
Ah, OK. Yes, I think we should just pick the last value. That's the same as what you'd get if instead of having consecutive runs in different files you just did all the runs with `restart append`, so I think it makes sense.

**Comment by TomNicholas** [comment]
> Ah, OK. Yes, I think we should just pick the last value.

Pick the last value just in `xbout.collect()`? Or also in `xbout.open_boutdataset()`?

**Comment by johnomotani** [comment]
> Pick the last value just in xbout.collect()? Or also in xbout.open_boutdataset()?

I'd pick the last value in `open_boutdataset` because otherwise we'd end up with an extra dimension with a funny length (number of files in t-dimension rather than number of timesteps).

**Comment by TomNicholas** [comment]
> I'd pick the last value in open_boutdataset because otherwise we'd end up with an extra dimension with a funny length (number of files in t-dimension rather than number of timesteps).

That's not what happens, explicitly specifying 'iteration' to be concatenated (by using the `data_vars` argument to `open_mfdataset`) means it gets returns as a 1D array along `'t'`.

However, having actually just tried it on real data I realise now that `iteration` doesn't mean what I thought it meant... I thought it meant "you've restarted this simulation ITERATION times".

I've changed it now but maybe I could have done all that in a much simpler way...

File: xbout/load.py [hunk 2]
```diff
@@ -91,33 +92,44 @@ def open_boutdataset(datapath='./BOUT.dmp.*.nc', inputfilepath=None,
...
+        # Store as ints because netCDF doesn't support bools, so we can't save
+        # bool attributes
+        metadata['keep_xboundaries'] = int(keep_xboundaries)
+        metadata['keep_yboundaries'] = int(keep_yboundaries)
+        ds = _set_attrs_on_all_vars(ds, 'metadata', metadata)
+
+        for var in _BOUT_TIME_DEPENDENT_META_VARS:
+            if var in ds:
+                # Assume different processors in x & y have same iteration etc.
+                latest_top_left = {dim: 0 for dim in ds[var].dims}
```
Comment (johnomotani) [inline xbout/load.py]: Doesn't this select the earliest value (`t=0`, when `dim='t'`) of var rather than the latest?

File: xbout/load.py [hunk 3]
```diff
@@ -243,6 +255,10 @@ def collect(varname, xind=None, yind=None, zind=None, tind=None,
         # If BOUT Version is not saved in the dataset
         version = 0
 
+    # Collect just stores iteration from BOUT.dmp.0 as a scalar
+    latest_top_left = {dim: 0 for dim in ds['iteration'].dims}
+    ds['iteration'] = ds['iteration'].isel(latest_top_left).squeeze(drop=True)
```
Comment (johnomotani) [inline xbout/load.py]: Is this needed if `iteration` is handled in `open_boutdataset`?

File: xbout/load.py (hunk 2 above)
Comment (TomNicholas) [reply xbout/load.py]: Yep, you're right

File: xbout/load.py (hunk 3 above)
Comment (TomNicholas) [reply xbout/load.py]: No, you're right

**Pull request merged**